File: 1396-design-underground-system/1396-design-underground-system.py
class UndergroundSystem:

    def __init__(self):
        self.travelTimes = {}
        self.checkIns = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkIns[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation, checkInTime = self.checkIns.pop(id)
        travel = (startStation, stationName)
        travelTime = t - checkInTime
        if travel in self.travelTimes:
            totalTime, count = self.travelTimes[travel]
            self.travelTimes[travel] = (totalTime+travelTime, count+1)
        else:
            self.travelTimes[travel] = (travelTime, 1)

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        travel = (startStation, endStation)
        totalTime, count = self.travelTimes[travel]
        return totalTime / count


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)

File: 1396-design-underground-system/1396-design-underground-system.java
class UndergroundSystem {
    Map<Integer, CheckInInfo> checkIns;
    Map<String, TravelInfo> travelTimes;
    public UndergroundSystem() {
        checkIns = new HashMap<>();
        travelTimes = new HashMap<>();
    }
    public void checkIn(int id, String stationName, int t) {
        checkIns.put(id, new CheckInInfo(stationName, t));
    }
    public void checkOut(int id, String stationName, int t) {
        CheckInInfo checkInfo = checkIns.remove(id);
        String travel = checkInfo.name + "," + stationName;
        int travelTime = t - checkInfo.time;
        if(travelTimes.containsKey(travel)){
            TravelInfo travelInfo = travelTimes.get(travel);
            travelInfo.totalTime += travelTime;
            travelInfo.count++;
        }
        else{
            travelTimes.put(travel, new TravelInfo(travelTime, 1));
        }
    }
    public double getAverageTime(String startStation, String endStation) {
        String travel = startStation+","+endStation;
        TravelInfo info = travelTimes.get(travel);
        return (double) info.totalTime / info.count;
    }
    class CheckInInfo{
        String name;
        int time;
        CheckInInfo(String name, int time){
            this.name = name;
            this.time = time;
        }
    }
    class TravelInfo{
        int totalTime;
        int count;
        TravelInfo(int totalTime, int count){
            this.totalTime = totalTime;
            this.count = count;
        }
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */